composer.json

```diff
@@ -60,11 +60,11 @@
         "yoast/phpunit-polyfills": "^1.0",
         "phpspec/prophecy": "^1.10.3",
         "squizlabs/php_codesniffer": "3.*",
-        "phpdocumentor/reflection": "^3.0||^4.0",
-        "symfony/console": "^3.0",
+        "phpdocumentor/reflection": "^3.0||^4.0||^5.x-dev",
+        "symfony/console": "^3.3||^4.4||^5.4||^6.2",
         "erusev/parsedown": "^1.6",
         "vierbergenlars/php-semver": "^3.0",
-        "symfony/lock": "3.3.x-dev#1ba6ac9",
+        "symfony/lock": "^3.3||^4.4||^5.4||^6.2",
```

Comment:
It would be best to avoid any/all changes to `composer.json`. Yes, this means that monolog 3.x won't be available to the monorepo (e.g. `google/cloud`), but as long as it's available to `google/cloud-core`, then the individual packages will be installable, which is the important thing.

Reply:
That's good news, but I think the docs generation GitHub Workflow will then break, because it uses the root `composer.json` 😕

But I hope I'm wrong about this 🤞🏻

Reply:
@bshaffer Maybe I'm missing something, but without support for Monolog 3, `google/cloud` is not compatible with Symfony 6. I would like to use the Google Vision API in my Symfony 6 application but can't unless this is supported.

Reply:
But I have to add... it makes local development on the root project as a whole more complicated when the dependencies on the root project are not updated. Do you work on each component separately in a sandboxed IDE?

When setting up the project for my PRs, I used https://github.com/googleapis/google-cloud-php/blob/main/CONTRIBUTING.md as the guideline…

Reply:
Could you please confirm once more that I should leave the root `composer.json` untouched? At the moment it just™ widens the requirements, which (I think) shouldn't have negative effects.

Reply:
@jamiehollern see https://github.com/googleapis/google-cloud-php/issues/5273 the root package `google/cloud` has been discouraged from use for years, and soon will be turned down entirely. You can use the Google Vision API in your application by installing [`google/cloud-vision`](https://github.com/googleapis/google-cloud-php-vision).

We don't want to support all the additional dependencies implied by this PR, and our docs generation is quite brittle and may have adverse consequences updating to a new major (and dev) version of `phpdocumentor`. 

I am happy to look into a potentially better configuration of dependencies in the future, after we remove the documentation component which is coming soon. But for now, it would be best to leave it out, so we can get this PR merged sooner rather than later.

Reply:
@bshaffer I'm a little bit in a deadlock here… 

* `google/cloud-logging` requires a released `google/cloud-core` version
* The currently released `google/cloud-core` supports `"monolog/monolog": "^1.0|^2.0"`, which prevents `google/cloud-logging` from installing Monolog V3
* Widening and releasing the requirement in `google/cloud` would break `google/cloud` for anyone using PHP 8.1 until `google/cloud-logging` is updated and released (and projects using `google/cloud-*` have updated all their dependencies

I understand the concern about not wanting to maintain the dependencies in the root `composer.json`, but I'm not sure how we can proceed here:

* The individual package tests will fail or at least not cover the Monolog V3 changes with tests
* If we don't update the root `composer.json`, we can't test and release both components simultaneously because the new features won't get picked up
* If we do update the root `composer.json`, there's the maintenance burden, but the individual package tests will still fail

A solution could be to move everything log-related from https://github.com/googleapis/google-cloud-php/tree/main/Core/src/Logger as well as the Monolog dependency into the Logging component and add Monolog V3 Support afterward, but I'm pretty sure this would break BC.

I'd appreciate your advice 🙏! If there's a chance to discuss this more directly/synchronously instead of in long-winded PR comment threads, you can find my contact details on my GH profile page!

Reply:
Then again, our previous conversation on this topic has brought us even more comments 😂 

<img width="780" alt="CleanShot 2022-10-13 at 00 15 20@2x" src="https://user-images.githubusercontent.com/67554/195458198-7c1fbf46-cdcd-4c12-a943-32d959403572.png">

Reply:
I do not think we are understanding each other @jamiehollern. I am not incorrect, let me try to explain.

When you run `composer install google/cloud-vision`, the `composer.json` for this repository is not effected. Composer is looking at [this file](https://github.com/googleapis/google-cloud-php/blob/main/Vision/composer.json), which uses `google/cloud-core` as a dependency, and that is determined by [this file](https://github.com/googleapis/google-cloud-php/blob/main/Core/composer.json). 

So the conflict you are getting is absolutely as a result of this repository, and this PR here can potentially fix it. But the root composer.json ([this file](https://github.com/googleapis/google-cloud-php/blob/main/composer.json)) and the package `google/cloud` (which installs this entire repository) is not involved.

@jeromegamez I see your point. Thank you for explaining this. I think I know how to fix the issue, but it will need to be in another PR. Basically, we'll need to configure composer to install `google/cloud-core` from the `Core` directory instead of from packagist. This is something I've done before.

Reply:
@bshaffer That's good news, thank you!

Another PR is even good, that gives me more time to clean up this one and move all the changes regarding better PHP 8.1 support and smaller fixes not related to Core and Logging to new PRs 💪🏻

Reply:
@jeromegamez try this https://github.com/googleapis/google-cloud-php/commit/2e0f7101e026bad55a618efca008e4516ebb9273

Reply:
Thanks! I had to adapt the script a little bit (https://github.com/jeromegamez/google-cloud-php/blob/monolog-3/.github/run-package-tests.sh) because using the local path didn't work unless setting `"canonical": false` ([Composer Repository Priorities](https://getcomposer.org/doc/articles/repository-priorities.md#making-repositories-non-canonical)).

The package tests are now executed, but the Monolog Upgrade is not covered because the PHP Version is fixed to 8.0 - fixing it to 8.1 would render the previous Monolog Versions untested.

I think the way to go would be to include the package tests in the test matrix, but this could get expensive (I'm thinking time and Packagist rate limits 🤔 )

Reply:
I've updated it now to be a little smarter (it uses repo options to set a version, which gets around the problem you mentioned)

Reply:
Thank you, this looks much better and the tests almost pass - the one error I see is left is https://github.com/googleapis/google-cloud-php/actions/runs/3246076562/jobs/5324410527#step:4:1480

```
1) Google\Cloud\Spanner\Tests\Snippet\Batch\BatchClientTest::testPubSubExample
Prophecy\Exception\Call\UnexpectedCallException: Unexpected method call on Double\ConnectionInterface\P7:
  - publishMessage(
        ["topic" => "...", "messages" => [["attributes" => ["snapshot" => xxx"]]]]
    )
expected calls were:
  - publishMessage(
        [..., exact("messages") => exact([["attributes" => ["snapshot" => "xxx"]]]), ...]
    )
  - pull(
        *
    )
```

I looked into it, but I'm not very versed in Prophecy, so I'm not (yet) sure how to fix it.

Reply:
@bshaffer I pulled in your latest changes from #5542 and reverted all changes in the root `composer.json` file. Thank you for your work on this; I couldn't have done it myself 😅 🙏